File: src/main/java/graph/Dijkstra.java
package graph;

import java.util.*;

public class Dijkstra {
    public static void main(String[] args) {
        System.out.println(new Dijkstra().calculateShortestPath());
    }


    private int calculateShortestPath() {
        Map<String, Set<NodeData>> graph = new HashMap<>();
        graph.put("1", Set.of(new NodeData("2", 2),
                new NodeData("4", 10),
                new NodeData("8", 1)));
        graph.put("2", Set.of(new NodeData("3", 3)));
        graph.put("3", Set.of(new NodeData("4", 4)));
        graph.put("4", Set.of(new NodeData("6", 6)));
        graph.put("5", Set.of(new NodeData("6", 6)));
        graph.put("6", Set.of(new NodeData("4", 4),
                new NodeData("5", 5),
                new NodeData("7", 7),
                new NodeData("8", 8)));
        graph.put("7", Set.of(new NodeData("6", 6)));
        graph.put("8", Set.of(new NodeData("6", 6),
                new NodeData("4", 1)));

        int shortestPath = calculateShortestPath("1", "4", graph);
        return shortestPath;
    }

    private int calculateShortestPath(String src, String dst, Map<String, Set<NodeData>> graph) {
        PriorityQueue<PathData> queue = new PriorityQueue<>(
                Comparator.comparingInt(PathData::getCost));
        Map<String, PathData> costMap = new HashMap<>();

        Set<String> visitedSet = new HashSet<>();

        PathData srcData = new PathData(src, 0, src);
        queue.add(srcData);
        costMap.put(src, srcData);

        while (!queue.isEmpty()) {
            PathData current = queue.poll();
            if (dst.equals(current.node)) {
                return current.cost;
            } else if (!visitedSet.contains(current.node)) {
                visitedSet.add(current.node);
                for (NodeData neighbor : graph.get(current.node)) {
                    if (costMap.containsKey(neighbor.node)) {
                        PathData neighborData = costMap.get(neighbor.node);
                        if (neighborData.cost > neighbor.weight + current.cost) {
                            queue.remove(neighborData);

                            neighborData.setCost(neighbor.weight + current.cost);
                            neighborData.setFrom(current.from);

                            queue.add(neighborData);
                        }
                    } else {
                        PathData neighborData = new PathData(neighbor.node, neighbor.weight + current.cost, current.from);
                        costMap.put(neighbor.node, neighborData);
                        queue.add(neighborData);
                    }
                }
            }
        }

        return Integer.MAX_VALUE;
    }

    class PathData {
        final String node;
        String from;

        Integer cost;

        PathData(String node, int cost, String from) {
            this.node = node;
            this.cost = cost;
            this.from = from;
        }

        public String getFrom() {
            return from;
        }

        public void setFrom(String from) {
            this.from = from;
        }

        public Integer getCost() {
            return cost;
        }

        public void setCost(Integer cost) {
            this.cost = cost;
        }
    }

    class NodeData {
        final String node;
        final int weight;

        NodeData(String node, int weight) {
            this.node = node;
            this.weight = weight;
        }
    }


}

File: src/main/kotlin/graph/Dijkstra.kt
package graph

import java.util.*

class Dijkstra {
    private fun calculateShortestPath(): Int {
        val graph: MutableMap<String, Set<NodeData>> =
            HashMap()
        graph["1"] = mutableSetOf(
            NodeData("2", 2),
            NodeData("4", 10),
            NodeData("8", 1)
        )
        graph["2"] = mutableSetOf(NodeData("3", 3))
        graph["3"] = mutableSetOf(NodeData("4", 4))
        graph["4"] = mutableSetOf(NodeData("6", 6))
        graph["5"] = mutableSetOf(NodeData("6", 6))
        graph["6"] = mutableSetOf(
            NodeData("4", 4),
            NodeData("5", 5),
            NodeData("7", 7),
            NodeData("8", 8)
        )
        graph["7"] = mutableSetOf(NodeData("6", 6))
        graph["8"] = mutableSetOf(
            NodeData("6", 6),
            NodeData("4", 1)
        )
        return calculateShortestPath("1", "4", graph)
    }

    private fun calculateShortestPath(src: String, dst: String, graph: Map<String, Set<NodeData>>): Int {
        val queue = PriorityQueue(
            Comparator.comparingInt { obj: PathData -> obj.cost })
        val costMap: MutableMap<String, PathData> = HashMap()
        val visitedSet: MutableSet<String> = HashSet()
        val srcData: PathData = PathData(src, 0, src)
        queue.add(srcData)
        costMap[src] = srcData
        while (!queue.isEmpty()) {
            val current = queue.poll()
            if (dst == current!!.node) {
                return current.cost
            } else if (!visitedSet.contains(current.node)) {
                visitedSet.add(current.node)
                for (neighbor in graph[current.node]!!) {
                    if (costMap.containsKey(neighbor.node)) {
                        val neighborData = costMap[neighbor.node]
                        if (neighborData!!.cost > neighbor.weight + current.cost) {
                            queue.remove(neighborData)
                            neighborData.cost = neighbor.weight + current.cost
                            neighborData.from = current.from
                            queue.add(neighborData)
                        }
                    } else {
                        val neighborData: PathData =
                            PathData(neighbor.node, neighbor.weight + current.cost, current.from)
                        costMap[neighbor.node] = neighborData
                        queue.add(neighborData)
                    }
                }
            }
        }
        return Int.MAX_VALUE
    }

    data class PathData(val node: String, var cost: Int, var from: String)
    data class NodeData(val node: String, val weight: Int)
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            Dijkstra().calculateShortestPath()

        }
    }
}